secant/Features/BackupFlow/RecoveryPhraseValidation.swift

```diff
@@ -0,0 +1,229 @@
+
+typealias RecoveryPhraseValidationStore = Store<RecoveryPhraseValidationState, RecoveryPhraseValidationAction>
+typealias RecoveryPhraseValidationViewStore = ViewStore<RecoveryPhraseValidationState, RecoveryPhraseValidationAction>
+
+struct RecoveryPhraseEnvironment {
+    var mainQueue: AnySchedulerOf<DispatchQueue>
+    var newPhrase: () -> Effect<RecoveryPhrase, AppError>
+}
+
+/// Represents the completion of a group of recovery words by de addition of one word into the given group
+struct RecoveryPhraseStepCompletion: Equatable {
+    var groupIndex: Int
+    var word: String
+}
+
+struct RecoveryPhraseValidationState: Equatable {
+    enum Step: Equatable {
+        case initial
+        case incomplete
+        case complete
+        case valid
+        case invalid
+    }
+
+    static let wordGroupSize = 6
+    static let phraseChunks = 4
+
+    var phrase: RecoveryPhrase
+    var missingIndices: [Int]
+    var missingWordChips: [PhraseChip.Kind]
+    var completion: [RecoveryPhraseStepCompletion]
+
+    var step: Step {
+        guard !completion.isEmpty else {
+            return  .initial
+        }
+
+        guard completion.count >= missingIndices.count else {
+            return .incomplete
+        }
+
+        return .complete
+    }
+
+    var isValid: Bool {
+        Self.resultingPhrase(from: completion, missingIndices: missingIndices, originalPhrase: phrase, numberOfGroups: missingIndices.count) == phrase.words
+    }
+}
+
+
+extension RecoveryPhraseValidationState {
+    static func initial(
+        phrase: RecoveryPhrase,
+        missingIndices: [Int],
+        missingWordsChips: [PhraseChip.Kind]
+    ) -> Self {
+        RecoveryPhraseValidationState(
+            phrase: phrase,
+            missingIndices: missingIndices,
+            missingWordChips: missingWordsChips,
+            completion: []
+        )
+    }
+
+    /// creates an initial `RecoveryPhraseValidationState` with no completions and random missing indices.
+    /// - Note: Use this function to create a random validation puzzle for a given phrase.
+    static func random(phrase: RecoveryPhrase) -> Self {
+        let missingIndices = Self.randomIndices()
+        let missingWordChipKind = Self.pickWordsFromMissingIndices(indices: missingIndices, phrase: phrase).shuffled()
+        return RecoveryPhraseValidationState(
+            phrase: phrase,
+            missingIndices: missingIndices,
+            missingWordChips: missingWordChipKind,
+            completion: []
+        )
+    }
+}
+
+extension RecoveryPhraseValidationState {
+    /// Given an array of RecoveryPhraseStepCompletion, missing indices, original phrase and the number of groups it was split into,
+    /// assembly the resulting phrase. This comes up with the "proposed solution" for the recovery phrase validation challenge.
+    /// - returns:an array of String containing the recovery phrase words ordered by the original phrase order.
+    static func resultingPhrase(
```

Comment:
I think this would read a lot simpler if you just calculate this right in the computed property instead of making a static function that requires you to pass in most of the sate from the store state.  Does having this static function help you achieve something that I'm missing?

Reply:
Yeay, It does not make much sense to have it dangling there anymore.